### PR TITLE
IC-1951 Added End Of Service Report Flag to single sent referral responses

### PR DESCRIPTION
**Pull request by mahalrmoj**
## What does this pull request do?
A derived field which indicates to the user interface if a EoS Report is required. 

## What is the intent behind these changes?
To avoid duplicating the business rule in UI

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt [hunk 1]
```diff
@@ -64,4 +66,10 @@ class ReferralConcluder(
 
     return null
   }
+
+  private fun getNumberOfAttendedAppointments(referral: Referral): Int {
+    return referral.actionPlan?.let {
+      actionPlanRepository.countNumberOfAttendedSessions(it.id)
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt:72]: Can we rename this to `countSessionsWithAttendanceRecorded` (or similar), please? "Attended" in domain terminology is when people turn up to their appointment

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt [hunk 2]
```diff
@@ -35,13 +35,15 @@ class ReferralConcluder(
     }
   }
 
+  fun requiresEosr(referral: Referral): Boolean {
+    return getNumberOfAttendedAppointments(referral) > 0
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt:39]: And it's concluded, right?

**Comment by sldblog** [comment]
👍 💯 for exposing the business rules in the business API

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt (hunk 1 above)
Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt:72]: Will do

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt (hunk 2 above)
Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt:39]: yes

Comment (sldblog) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt]: Can we add those conditions?

Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt]: Added a checked that the referral has infact ended for some reason

**Review by rgforsyth — APPROVED** [review]

**Review by sldblog — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt [hunk 3]
```diff
@@ -81,22 +85,23 @@ class ReferralController(
 
     return ResponseEntity
       .created(location)
-      .body(SentReferralDTO.from(sentReferral))
+      .body(SentReferralDTO.from(sentReferral, referralConcluder.requiresEndOfServiceReportCreation(sentReferral)))
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/ReferralController.kt:88]: nit: i think it's safe to assume that freshly sent referrals don't require EOSRs, but it's fine to do this anyway.

**Review by tomsmyers — APPROVED** [review]

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt (hunk 2 above)
Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralConcluder.kt]: Agreed logic ... true if:

endOfServiceReport does not exist 
and
(allSessionsDelivered) || (endRequested && atLeastOneSessionDelivered)

**Pull request merged**